src/meta/src/rpc/server.rs

```diff
@@ -129,167 +132,327 @@ pub async fn rpc_serve(
     }
 }
 
-pub async fn register_leader_for_meta<S: MetaStore>(
-    addr: String,
-    meta_store: Arc<S>,
-    lease_time: u64,
-) -> MetaResult<(MetaLeaderInfo, JoinHandle<()>, Sender<()>)> {
-    let mut tick_interval = tokio::time::interval(Duration::from_secs(lease_time / 2));
-    loop {
-        tick_interval.tick().await;
-        let old_leader_info = match meta_store
-            .get_cf(META_CF_NAME, META_LEADER_KEY.as_bytes())
-            .await
-        {
-            Err(MetaStoreError::ItemNotFound(_)) => vec![],
-            Ok(v) => v,
-            _ => {
-                continue;
-            }
-        };
-        let old_leader_lease = match meta_store
-            .get_cf(META_CF_NAME, META_LEASE_KEY.as_bytes())
-            .await
-        {
-            Err(MetaStoreError::ItemNotFound(_)) => vec![],
-            Ok(v) => v,
-            _ => {
-                continue;
-            }
-        };
-        let now = SystemTime::now()
-            .duration_since(UNIX_EPOCH)
-            .expect("Time went backwards");
-        if !old_leader_lease.is_empty() {
-            let lease_info = MetaLeaseInfo::decode(&mut old_leader_lease.as_slice()).unwrap();
-
-            if lease_info.lease_expire_time > now.as_secs()
-                && lease_info.leader.as_ref().unwrap().node_address != addr
-            {
-                let err_info = format!(
-                    "the lease {:?} does not expire, now time: {}",
-                    lease_info,
-                    now.as_secs(),
-                );
-                tracing::error!("{}", err_info);
-                bail!(err_info);
-            }
+// get duration since epoch
+fn since_epoch() -> Duration {
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("Time went backwards")
+}
+
+struct ElectionResult {
+    pub meta_leader_info: MetaLeaderInfo,
+    pub meta_lease_info: MetaLeaseInfo,
+
+    // True if current node is leader
+    pub is_leader: bool,
+}
+
+// Sets a leader. Does not renew leaders term
+// lease_id: random id identifying the leader via the lease.
+async fn run_for_election<S: MetaStore>(
```

Comment:
How about moving leader election part to a completely new file?

Reply:
This could work, but almost the entire file is election related. I Will move it to a new file and keep a wrapper in `server.rs`.